Objects/object.c

```diff
@@ -2485,12 +2485,12 @@ new_reference(PyObject *op)
     op->ob_refcnt = 1;
 #endif
 #else
-    op->ob_tid = _Py_ThreadId();
+    _Py_atomic_store_uintptr_relaxed(&op->ob_tid, _Py_ThreadId());
     op->ob_flags = 0;
     op->ob_mutex = (PyMutex){ 0 };
-    op->ob_gc_bits = 0;
-    op->ob_ref_local = 1;
-    op->ob_ref_shared = 0;
+    _Py_atomic_store_uint8_relaxed(&op->ob_gc_bits, 0);
+    _Py_atomic_store_uint32_relaxed(&op->ob_ref_local, 1);
+    _Py_atomic_store_ssize_relaxed(&op->ob_ref_shared, 0);
```

Comment:
I think in this case we only want to use `_Py_atomic_store_[type]_relaxed` in `_Py_THREAD_SANITIZER` builds. For performance reasons, we want to use non-atomic stores so that some of the stores to adjacent fields get coalesced into a single larger store (`ob_flags`, `ob_mutex`, `ob_gc_bits` and `ob_ref_local`).

For example, see `_Py_IsOwnedByCurrentThread`:
https://github.com/python/cpython/blob/e41ec8e18b078024b02a742272e675ae39778536/Include/object.h#L251-L255

We don't do this generally, but `_Py_IsOwnedByCurrentThread()` and `new_reference()` are particularly performance sensitive.

Reply:
+